From 1d51a3bd368d99fd3baac1da8c5621ac3d1ec76a Mon Sep 17 00:00:00 2001
From: noctis <970308389@qq.com>
Date: Tue, 4 Aug 2026 00:51:30 +0800
Subject: [PATCH 1/2] llvm: enable the LoongArch backend on Android

SwiftShader's LLVM 16 already carries the LoongArch backend and Linux configuration, but its Android triples and generated Soong source lists omit it. Add the Android target configuration and teach the generators to retain it so Mesa can use LLVM-backed drivers on LoongArch64.

Change-Id: If0a0db01f72ee4670c33bbba372d9d7c6add6d72
---
 third_party/llvm-16.0/Android.bp              | 33 +++++++++++++++++++
 .../include/llvm/Config/AsmParsers.def        |  3 ++
 .../include/llvm/Config/AsmPrinters.def       |  3 ++
 .../include/llvm/Config/Disassemblers.def     |  3 ++
 .../android/include/llvm/Config/Targets.def   |  3 ++
 .../android/include/llvm/Config/llvm-config.h | 18 ++++++++++
 .../llvm-16.0/scripts/generate_build_files.py |  3 +-
 .../llvm-16.0/scripts/template_Android.bp     |  7 ++++
 third_party/llvm-16.0/scripts/update.py       |  1 +
 9 files changed, 73 insertions(+), 1 deletion(-)

diff --git a/third_party/llvm-16.0/Android.bp b/third_party/llvm-16.0/Android.bp
index aaf033182..488104200 100644
--- a/third_party/llvm-16.0/Android.bp
+++ b/third_party/llvm-16.0/Android.bp
@@ -1274,6 +1274,37 @@ cc_defaults {
                 "llvm/lib/Transforms/IPO/BarrierNoopPass.cpp",
             ],
         },
+        loongarch64: {
+            srcs: [
+                "llvm/lib/Target/LoongArch/AsmParser/LoongArchAsmParser.cpp",
+                "llvm/lib/Target/LoongArch/Disassembler/LoongArchDisassembler.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchAsmPrinter.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchExpandAtomicPseudoInsts.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchExpandPseudoInsts.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchFrameLowering.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchISelDAGToDAG.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchInstrInfo.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchMCInstLower.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchRegisterInfo.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchSubtarget.cpp",
+                "llvm/lib/Target/LoongArch/LoongArchTargetMachine.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchAsmBackend.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchELFObjectWriter.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchELFStreamer.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchInstPrinter.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchMCAsmInfo.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchMCCodeEmitter.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchMCExpr.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchMCTargetDesc.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchMatInt.cpp",
+                "llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchTargetStreamer.cpp",
+                "llvm/lib/Target/LoongArch/TargetInfo/LoongArchTargetInfo.cpp",
+                "llvm/lib/TargetParser/LoongArchTargetParser.cpp",
+                "llvm/lib/Transforms/IPO/BarrierNoopPass.cpp",
+            ],
+        },
         riscv64: {
             srcs: [
                 "llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp",
@@ -1502,9 +1533,11 @@ cc_defaults {
         "llvm/lib/Target/X86",
         "llvm/lib/Target/AArch64",
         "llvm/lib/Target/ARM",
+        "llvm/lib/Target/LoongArch",
         "llvm/lib/Target/RISCV",
         "configs/common/lib/Target/AArch64",
         "configs/common/lib/Target/ARM",
+        "configs/common/lib/Target/LoongArch",
         "configs/common/lib/Target/X86",
         "configs/common/lib/Target/RISCV",
         "configs/common/lib/ExecutionEngine/JITLink",
diff --git a/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmParsers.def b/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmParsers.def
index 808436cfc..3b3141b95 100644
--- a/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmParsers.def
+++ b/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmParsers.def
@@ -33,6 +33,9 @@ LLVM_ASM_PARSER(ARM)
 #if defined(__i386__) || defined(__x86_64__)
 LLVM_ASM_PARSER(X86)
 #endif
+#if defined(__loongarch__)
+LLVM_ASM_PARSER(LoongArch)
+#endif
 #if defined(__riscv)
 LLVM_ASM_PARSER(RISCV)
 #endif
diff --git a/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmPrinters.def b/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmPrinters.def
index 5566edc34..e4ae9e8fb 100644
--- a/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmPrinters.def
+++ b/third_party/llvm-16.0/configs/android/include/llvm/Config/AsmPrinters.def
@@ -33,6 +33,9 @@ LLVM_ASM_PRINTER(ARM)
 #if defined(__i386__) || defined(__x86_64__)
 LLVM_ASM_PRINTER(X86)
 #endif
+#if defined(__loongarch__)
+LLVM_ASM_PRINTER(LoongArch)
+#endif
 #if defined(__riscv)
 LLVM_ASM_PRINTER(RISCV)
 #endif
diff --git a/third_party/llvm-16.0/configs/android/include/llvm/Config/Disassemblers.def b/third_party/llvm-16.0/configs/android/include/llvm/Config/Disassemblers.def
index eb3ffa45e..ae13ad642 100644
--- a/third_party/llvm-16.0/configs/android/include/llvm/Config/Disassemblers.def
+++ b/third_party/llvm-16.0/configs/android/include/llvm/Config/Disassemblers.def
@@ -33,6 +33,9 @@ LLVM_DISASSEMBLER(ARM)
 #if defined(__i386__) || defined(__x86_64__)
 LLVM_DISASSEMBLER(X86)
 #endif
+#if defined(__loongarch__)
+LLVM_DISASSEMBLER(LoongArch)
+#endif
 #if defined(__riscv)
 LLVM_DISASSEMBLER(RISCV)
 #endif
diff --git a/third_party/llvm-16.0/configs/android/include/llvm/Config/Targets.def b/third_party/llvm-16.0/configs/android/include/llvm/Config/Targets.def
index ef9607fc8..3c827c91b 100644
--- a/third_party/llvm-16.0/configs/android/include/llvm/Config/Targets.def
+++ b/third_party/llvm-16.0/configs/android/include/llvm/Config/Targets.def
@@ -32,6 +32,9 @@ LLVM_TARGET(ARM)
 #if defined(__i386__) || defined(__x86_64__)
 LLVM_TARGET(X86)
 #endif
+#if defined(__loongarch__)
+LLVM_TARGET(LoongArch)
+#endif
 #if defined(__riscv)
 LLVM_TARGET(RISCV)
 #endif
diff --git a/third_party/llvm-16.0/configs/android/include/llvm/Config/llvm-config.h b/third_party/llvm-16.0/configs/android/include/llvm/Config/llvm-config.h
index 5be219d28..b4bbc1e8d 100644
--- a/third_party/llvm-16.0/configs/android/include/llvm/Config/llvm-config.h
+++ b/third_party/llvm-16.0/configs/android/include/llvm/Config/llvm-config.h
@@ -37,6 +37,8 @@
 #define LLVM_DEFAULT_TARGET_TRIPLE "armv7-linux-androideabi"
 #elif defined(__aarch64__)
 #define LLVM_DEFAULT_TARGET_TRIPLE "aarch64-linux-android"
+#elif defined(__loongarch__)
+#define LLVM_DEFAULT_TARGET_TRIPLE "loongarch64-linux-android"
 #elif defined(__riscv)
 #define LLVM_DEFAULT_TARGET_TRIPLE "riscv64-linux-android"
 #else
@@ -58,6 +60,8 @@
 #define LLVM_HOST_TRIPLE "armv7-linux-androideabi"
 #elif defined(__aarch64__)
 #define LLVM_HOST_TRIPLE "aarch64-linux-android"
+#elif defined(__loongarch__)
+#define LLVM_HOST_TRIPLE "loongarch64-linux-android"
 #elif defined(__riscv)
 #define LLVM_HOST_TRIPLE "riscv64-linux-android"
 #else
@@ -71,6 +75,8 @@
 #define LLVM_NATIVE_ARCH ARM
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_ARCH X86
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_ARCH LoongArch
 #elif defined(__mips__)
 #define LLVM_NATIVE_ARCH Mips
 #elif defined(__powerpc64__)
@@ -88,6 +94,8 @@
 #define LLVM_NATIVE_ASMPARSER LLVMInitializeARMAsmParser
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_ASMPARSER LLVMInitializeX86AsmParser
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_ASMPARSER LLVMInitializeLoongArchAsmParser
 #elif defined(__mips__)
 #define LLVM_NATIVE_ASMPARSER LLVMInitializeMipsAsmParser
 #elif defined(__powerpc64__)
@@ -105,6 +113,8 @@
 #define LLVM_NATIVE_ASMPRINTER LLVMInitializeARMAsmPrinter
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_ASMPRINTER LLVMInitializeX86AsmPrinter
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_ASMPRINTER LLVMInitializeLoongArchAsmPrinter
 #elif defined(__mips__)
 #define LLVM_NATIVE_ASMPRINTER LLVMInitializeMipsAsmPrinter
 #elif defined(__powerpc64__)
@@ -122,6 +132,8 @@
 #define LLVM_NATIVE_DISASSEMBLER LLVMInitializeARMDisassembler
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_DISASSEMBLER LLVMInitializeX86Disassembler
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_DISASSEMBLER LLVMInitializeLoongArchDisassembler
 #elif defined(__mips__)
 #define LLVM_NATIVE_DISASSEMBLER LLVMInitializeMipsDisassembler
 #elif defined(__powerpc64__)
@@ -139,6 +151,8 @@
 #define LLVM_NATIVE_TARGET LLVMInitializeARMTarget
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_TARGET LLVMInitializeX86Target
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_TARGET LLVMInitializeLoongArchTarget
 #elif defined(__mips__)
 #define LLVM_NATIVE_TARGET LLVMInitializeMipsTarget
 #elif defined(__powerpc64__)
@@ -156,6 +170,8 @@
 #define LLVM_NATIVE_TARGETINFO LLVMInitializeARMTargetInfo
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_TARGETINFO LLVMInitializeX86TargetInfo
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_TARGETINFO LLVMInitializeLoongArchTargetInfo
 #elif defined(__mips__)
 #define LLVM_NATIVE_TARGETINFO LLVMInitializeMipsTargetInfo
 #elif defined(__powerpc64__)
@@ -173,6 +189,8 @@
 #define LLVM_NATIVE_TARGETMC LLVMInitializeARMTargetMC
 #elif defined(__i386__) || defined(__x86_64__)
 #define LLVM_NATIVE_TARGETMC LLVMInitializeX86TargetMC
+#elif defined(__loongarch__)
+#define LLVM_NATIVE_TARGETMC LLVMInitializeLoongArchTargetMC
 #elif defined(__mips__)
 #define LLVM_NATIVE_TARGETMC LLVMInitializeMipsTargetMC
 #elif defined(__powerpc64__)
diff --git a/third_party/llvm-16.0/scripts/generate_build_files.py b/third_party/llvm-16.0/scripts/generate_build_files.py
index 4d182e680..1e5ceaf4d 100755
--- a/third_party/llvm-16.0/scripts/generate_build_files.py
+++ b/third_party/llvm-16.0/scripts/generate_build_files.py
@@ -431,6 +431,7 @@ android_bp_template_data = {
     'files_x86' : format_file_list_for_android_bp(files_x86, indent=4),
     'files_AArch64' : format_file_list_for_android_bp(files_AArch64, indent=4),
     'files_ARM' : format_file_list_for_android_bp(files_ARM, indent=4),
+    'files_LoongArch' : format_file_list_for_android_bp(files_LoongArch, indent=4),
     'files_RISCV' : format_file_list_for_android_bp(files_RISCV, indent=4),
 }
 with open(ANDROID_BP_TEMPLATE_PATH, 'r') as f:
@@ -514,4 +515,4 @@ with open(BUILD_GN_TEMPLATE_PATH, 'r') as f:
     build_gn_template = CustomTemplate(f.read())
     result = build_gn_template.substitute(build_gn_template_data)
     cmake_output = open(DESTINATION_BUILD_GN_PATH, "w")
-    cmake_output.write(result)
\ No newline at end of file
+    cmake_output.write(result)
diff --git a/third_party/llvm-16.0/scripts/template_Android.bp b/third_party/llvm-16.0/scripts/template_Android.bp
index 00b674bba..0a97fb650 100644
--- a/third_party/llvm-16.0/scripts/template_Android.bp
+++ b/third_party/llvm-16.0/scripts/template_Android.bp
@@ -34,6 +34,11 @@ cc_defaults {
 %$%files_AArch64
             ],
         },
+        loongarch64: {
+            srcs: [
+%$%files_LoongArch
+            ],
+        },
         riscv64: {
             srcs: [
 %$%files_RISCV
@@ -56,9 +61,11 @@ cc_defaults {
         "llvm/lib/Target/X86",
         "llvm/lib/Target/AArch64",
         "llvm/lib/Target/ARM",
+        "llvm/lib/Target/LoongArch",
         "llvm/lib/Target/RISCV",
         "configs/common/lib/Target/AArch64",
         "configs/common/lib/Target/ARM",
+        "configs/common/lib/Target/LoongArch",
         "configs/common/lib/Target/X86",
         "configs/common/lib/Target/RISCV",
         "configs/common/lib/ExecutionEngine/JITLink",
diff --git a/third_party/llvm-16.0/scripts/update.py b/third_party/llvm-16.0/scripts/update.py
index 7c3641a70..b0084394c 100644
--- a/third_party/llvm-16.0/scripts/update.py
+++ b/third_party/llvm-16.0/scripts/update.py
@@ -57,6 +57,7 @@ LLVM_TRIPLES = {
         ('__i386__', 'i686-linux-android'),
         ('__arm__', 'armv7-linux-androideabi'),
         ('__aarch64__', 'aarch64-linux-android'),
+        ('__loongarch__', 'loongarch64-linux-android'),
         ('__riscv', 'riscv64-linux-android'),
     ],
     'linux': [
